# **Mini Project:**
# **Predicting Airline Passenger Satisfaction**

# **1.0 Business Understanding**
---
## 1.1 Business Overview
We are trying to analyse the passenger satisfaction in the 


## 1.2 Stakeholder 
- Airline Companies
- Airline Investors
- Airline Staff

## 1.3 Problem Statement 


## 1.4 Objectives 
### 1.4.1 Main Objective 


### 1.4.2 Specific Objectives 


## 1.5 Research Questions 
1. 
   
## 1.6 Success Criteria 


# 2.0 Data Understanding 

### 2.1 Importing the Required Libraries

In [1]:
# importing the libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split, GridSearchCV
from imblearn.over_sampling import SMOTE
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import LinearSVC
from sklearn.metrics import ConfusionMatrixDisplay

### 2.2 Loading the Dataset

In [2]:
airline_df = pd.read_csv("dataset.csv")
airline_df.head()

,id,Gender,Customer Type,Age,Type of Travel,Class,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,...,Inflight entertainment,On-board service,Leg room service,Baggage handling,Checkin service,Inflight service,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes,satisfaction
0,70172,Male,Loyal Customer,13,Personal Travel,Eco Plus,460,3,4,3,...,5,4,3,4,4,5,5,25,18.0,neutral or dissatisfied
1,5047,Male,disloyal Customer,25,Business travel,Business,235,3,2,3,...,1,1,5,3,1,4,1,1,6.0,neutral or dissatisfied
2,110028,Female,Loyal Customer,26,Business travel,Business,1142,2,2,2,...,5,4,3,4,4,4,5,0,0.0,satisfied
3,24026,Female,Loyal Customer,25,Business travel,Business,562,2,5,5,...,2,2,5,3,1,4,2,11,9.0,neutral or dissatisfied
4,119299,Male,Loyal Customer,61,Business travel,Business,214,3,3,3,...,3,3,4,4,3,3,3,0,0.0,satisfied


### 2.3 Dataset Overview

In [3]:
# Dataset information 

dataset_info = pd.DataFrame({
    'Metric': ['Passenger Responses', 'Features', 'Missing Values', 'Duplicates'],
    'Value': [
        airline_df.shape[0],
        airline_df.shape[1],
        airline_df.isnull().sum().sum(),
        airline_df.duplicated().sum()
    ]
})
dataset_info

,Metric,Value
0,Passenger Responses,129880
1,Features,24
2,Missing Values,393
3,Duplicates,0


From the summary above:
- The dataset contains 129,880 passenger responses across 24 features.
- The initial data quality assessment shows 393 missing values across the dataset and no duplicate records.

The absence of duplicate rows suggests that each observation represents a unique passenger survey response. However, the presence of missing values indicates that some variables may require preprocessing, such as imputation or removal of the rows, before modelling can be performed. This step will be important to ensure that machine learning models can be trained without bias or loss of information.

In [4]:
# Dataset datatypes 

airline_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129880 entries, 0 to 129879
Data columns (total 24 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   id                                 129880 non-null  int64  
 1   Gender                             129880 non-null  object 
 2   Customer Type                      129880 non-null  object 
 3   Age                                129880 non-null  int64  
 4   Type of Travel                     129880 non-null  object 
 5   Class                              129880 non-null  object 
 6   Flight Distance                    129880 non-null  int64  
 7   Inflight wifi service              129880 non-null  int64  
 8   Departure/Arrival time convenient  129880 non-null  int64  
 9   Ease of Online booking             129880 non-null  int64  
 10  Gate location                      129880 non-null  int64  
 11  Food and drink                     1298

In [5]:
# Understanding the columns with the missing values

airline_df.isnull().sum().sort_values(ascending = True) 

id                                     0
Departure Delay in Minutes             0
Cleanliness                            0
Inflight service                       0
Checkin service                        0
Baggage handling                       0
Leg room service                       0
On-board service                       0
Inflight entertainment                 0
Seat comfort                           0
Online boarding                        0
Food and drink                         0
Gate location                          0
Ease of Online booking                 0
Departure/Arrival time convenient      0
Inflight wifi service                  0
Flight Distance                        0
Class                                  0
Type of Travel                         0
Age                                    0
Customer Type                          0
Gender                                 0
satisfaction                           0
Arrival Delay in Minutes             393
dtype: int64

In [6]:
# Dropping the rows with the missing values 

airline_df = airline_df.dropna()
airline_df.isnull().sum().sort_values(ascending = True)

id                                   0
Departure Delay in Minutes           0
Cleanliness                          0
Inflight service                     0
Checkin service                      0
Baggage handling                     0
Leg room service                     0
On-board service                     0
Inflight entertainment               0
Seat comfort                         0
Online boarding                      0
Food and drink                       0
Gate location                        0
Ease of Online booking               0
Departure/Arrival time convenient    0
Inflight wifi service                0
Flight Distance                      0
Class                                0
Type of Travel                       0
Age                                  0
Customer Type                        0
Gender                               0
Arrival Delay in Minutes             0
satisfaction                         0
dtype: int64

In [7]:
# Number of unique values in the dataset per column

airline_df.nunique()

id                                   129487
Gender                                    2
Customer Type                             2
Age                                      75
Type of Travel                            2
Class                                     3
Flight Distance                        3821
Inflight wifi service                     6
Departure/Arrival time convenient         6
Ease of Online booking                    6
Gate location                             6
Food and drink                            6
Online boarding                           6
Seat comfort                              6
Inflight entertainment                    6
On-board service                          6
Leg room service                          6
Baggage handling                          5
Checkin service                           6
Inflight service                          6
Cleanliness                               6
Departure Delay in Minutes              464
Arrival Delay in Minutes        

In [8]:
airline_df = airline_df.drop("id", axis = 'columns')
airline_df.head(2)

,Gender,Customer Type,Age,Type of Travel,Class,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,Gate location,...,Inflight entertainment,On-board service,Leg room service,Baggage handling,Checkin service,Inflight service,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes,satisfaction
0,Male,Loyal Customer,13,Personal Travel,Eco Plus,460,3,4,3,1,...,5,4,3,4,4,5,5,25,18.0,neutral or dissatisfied
1,Male,disloyal Customer,25,Business travel,Business,235,3,2,3,3,...,1,1,5,3,1,4,1,1,6.0,neutral or dissatisfied


#### Missing Values Analysis
- An inspection of missing values shows that only one feature, Arrival Delay in Minutes, contains missing responses. Specifically, 393 values are missing out of a total of 129,880, representing approximately 0.3% of the dataset.
- Since the number of missing values is negligible and limited to a single variable, the rows containing missing values were removed. This implifies preprocessing on the overall dataset size.

#### Unique Values, Data Types and Feature Inspection
- The number of unique values for each column was then examined to better understand the structure of the dataset. Several variables, such as Gender, Customer Type, Type of Travel, and Class, contain a small number of unique values, which is consistent with their categorical data type.
- In contrast, variables such as Age, Flight Distance, and the delay variables contain many unique values, which goes hand in hand that they represent continuous or numerical features according to their data type.
- Additionally, many service-related variables; Seat comfort, Inflight entertainment, and Online boarding contain a limited set of integer values, suggesting that they represent ordinal ratings provided by passengers in the survey
  
#### Removing the id Column
- The id column was removed from the dataset since it serves only as a unique identifier for each passenger response and does not contain meaningful information for predicting passenger satisfaction. 

### 2.4 Summary Statistics 

In [12]:
airline_df.describe()

,Age,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,Gate location,Food and drink,Online boarding,Seat comfort,Inflight entertainment,On-board service,Leg room service,Baggage handling,Checkin service,Inflight service,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes
count,129487.000000,129487.000000,129487.000000,129487.000000,129487.000000,129487.000000,129487.000000,129487.000000,129487.000000,129487.000000,129487.000000,129487.000000,129487.000000,129487.000000,129487.000000,129487.000000,129487.000000,129487.000000
mean,39.428761,1190.210662,2.728544,3.057349,2.756786,2.976909,3.204685,3.252720,3.441589,3.358067,3.383204,3.351078,3.631886,3.306239,3.642373,3.286222,14.643385,15.091129
std,15.117597,997.560954,1.329235,1.526787,1.401662,1.278506,1.329905,1.350651,1.319168,1.334149,1.287032,1.316132,1.180082,1.266146,1.176614,1.313624,37.932867,38.465650
min,7.000000,31.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,27.000000,414.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,3.000000,3.000000,3.000000,2.000000,0.000000,0.000000
50%,40.000000,844.000000,3.000000,3.000000,3.000000,3.000000,3.000000,3.000000,4.000000,4.000000,4.000000,4.000000,4.000000,3.000000,4.000000,3.000000,0.000000,0.000000
75%,51.000000,1744.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000,5.000000,4.000000,4.000000,4.000000,5.000000,4.000000,5.000000,4.000000,12.000000,13.000000
max,85.000000,4983.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,1592.000000,1584.000000


# 3.0 Exploratory Data Analysis

### 3.1 Target Variable Distribution

### 3.2 Numerical Feature Distributions

### 3.3 Categorical Feature Analysis

### 3.4 Correlation Analysis

### 3.5 Outlier Detection

# 3.0 Data Preparation 

In [ ]:
corr_matrix = airline_df.corr(numeric_only=True)
corr_matrix

# 4.0 Data Preprocessing 

In [ ]:
plt.figure(figsize = (12,8))
sns.heatmap(airline_df.corr(numeric_only = True), cmap="coolwarm")
plt.title("Correlation Heatmap")
plt.show()

# 5.0 Modelling 

# 6.0 Evaluation & Conclusion 